extensions/poetry/main.ts

```diff
@@ -0,0 +1,108 @@
+import { PhylumApi } from "phylum"
+import { parseDryRun } from "./parse.ts"
+import { red, green, yellow } from "https://deno.land/std@0.150.0/fmt/colors.ts";
+
+class FileBackup {
+  readonly fileName: string
+  readonly fileContent: string | null
+
+  constructor(fileName: string) {
+    this.fileName = fileName
+    this.fileContent = null
+  }
+
+  async backup() {
+    try {
+      this.fileContent = await Deno.readTextFile(this.fileName)
+    } catch (e) { }
+  }
+
+  async restoreOrDelete() {
+    try {
+      if (this.fileContent != null) {
+        await Deno.writeTextFile(this.fileName, this.fileContent)
+      } else {
+        await Deno.remove(this.fileName)
+      }
+    } catch (e) { }
+  }
+}
+
+// Parse the output of `poetry` subcommands that support the `--dry-run` flag,
+// and submit the packages.
+//
+// Add the `-vvv` flags to get a detailed report of the dependency resolution
+// process. This way, we can track the actual lockfile changes, as specifying
+// `--dry-run` only would not output anything in combination with flags such
+// as `--lock` that do not perform the actual operations.
+async function poetryCheckDryRun(subcommand: string, args: string[]): boolean {
+  // Read and backup the current poetry lockfile contents.
+  const lockfileBackup = new FileBackup('poetry.lock')
+  const manifestBackup = new FileBackup('pyproject.toml')
+
+  await lockfileBackup.backup()
+  await manifestBackup.backup()
+
+  let process = Deno.run({
+    cmd: ['poetry', subcommand, '-vvv', '-n', '--dry-run', ...args.map(s => s.toString())],
+    stdout: 'piped',
+    stderr: 'piped',
+  })
+
+  await process.status()
+  await process.close()
+
+  // If it existed before, restore the previous contents of the lockfile;
+  // otherwise, delete it. This is a workaround to the fact that in poetry
+  // 1.1.x, the `--dry-run` argument does not prevent the lockfile from 
+  // being modified. This is not fixed as of poetry 1.1.14.
+  // Prudently, do the same for the manifest (pyproject.toml).
```

Comment:
>   // Prudently, do the same for the manifest (pyproject.toml).

You do it just for fun? I'm not sure I understand the benefit of doing it for a file that is never affected.

Reply:
It is affected with at least one combination of parameters, though I don't remember exactly which one.

We cannot control Poetry's invariants or the way they change things over time, but we do know we want our first `poetry` command not to affect those two files specifically and backing both up guarantees to us that no undesired change will be ever made.

The cost of this code is approximately zero, and the benefit is a guarantee over something that is not under our control and about which we'd have otherwise to make assumptions about, crossing fingers that no PR gets into Poetry that violates those assumptions.

Reply:
> I'd say the cost of creating a file in the user's project directory is always **significantly** bigger than zero. 

We don't do that, though? We store the file's contents in memory, run `poetry`, and then immediately either restore the contents of the file (if the file was present before) or delete the file (if it wasn't). The only destructive action is thus writing data to a file that was read from the same file, and the action is limited to the files we have requested permissions for. The failure cases are: the user having their files changed by Poetry (which we can't avoid anyway without a backup) because our extension has crashed before restoring the files, and OS-I/O level errors that corrupt the file (over which we will never have any control whatsoever).

> There's also a possibility with every file we touch that we will do irrecoverable damage to that file.

At some point I suggested enforcing a clean git tree / a `--force` flag, but that is way overkill for this use case and probably annoying to the user who wants to add a number of packages in multiple commands. What are some examples of failure cases in which we could do irrecoverable damage to a file?

Reply:
> We don't do that, though? We store the file's contents in memory, run poetry, and then immediately either restore the contents of the file (if the file was present before) or delete the file (if it wasn't). The only destructive action is thus writing data to a file that was read from the same file, and the action is limited to the files we have requested permissions for. The failure cases are: the user having their files changed by Poetry (which we can't avoid anyway without a backup) because our extension has crashed before restoring the files, and OS-I/O level errors that corrupt the file (over which we will never have any control whatsoever).

We do have control over corruption though, that's why I previously suggested to atomically swap the file out after writing to a new one.

Reply:
> We do have control over corruption though, that's why I previously suggested to atomically swap the file out after writing to a new one.

The concept is that a corruption can only happen in a way the extension cannot be held responsible for -- i.e. bad hard drive or another `poetry` command (or worse, another random one) writing to the file at the same time. Both _aren't_ avoided by writing to a new file and atomically renaming: in the first case, we just write bad data to a different file, and in the second case, we could get the file overwritten _after_ having moved it, which is also problematic at the outset because then we can't guarantee we uploaded the right packages.